### DISCUSSION TITLE
What do you think this function does?

### DISCUSSION THREAD

**Discussion by kabiri** [NationalSecurityAgency/ghidra#4303] [Q&A]
What do you think this function does?
Why does my code have all of these functions?

![Fn01](https://user-images.githubusercontent.com/8082245/171414873-31e9a73d-78c2-42d3-9cfd-b4472948cf7e.png)

**Comment by astrelsky** [comment]
https://en.wikipedia.org/wiki/Thunk

**Reply by kabiri** [reply]
I do not think it's to delay.
```
void FUN_000004b4(void)

{
  FUN_00001382(0);
  return;
}

undefined4 FUN_00001382(void)
{
  return 0;
}
```
But this is to delay.

```
void _zig_mybe_delayT(uint param_11)

{
  uint i;
  
  for (i = 0; i < param_11 >> 0x10; i = i + 1) {
    _zig_mybe_delay();
  }
  return;
}

void _zig_mybe_delay(void)

{
  uint i;
  
  for (i = 0; i < 0x400; i = i + 1) {
  }
  return;
}
```
The assembly code is for the microprocessor.

**Reply by astrelsky** [reply]
I've never seen a thunk used for a delay. Other than that, the wiki page is fairly accurate. I recommend reading it all the way through. You'll find the answer you're looking for.

**Reply by agatti** [reply]
+1 on thunk - looks like something a C/C++ compiler would generate too.

**Comment by kabiri** [comment]
what does it mean?

```
  for (uVar1 = 0; uVar1 < 0x100; uVar1 = uVar1 + 1) {
    *(undefined4 *)(uVar1 * 4 + -0x80000000) = *(undefined4 *)(uVar1 * 4);
  }
  FUN_00001420(0x80000000);
```

**Reply by Wall-AF** [reply]
Copy the first 256 4-byte elements or 1K bytes (starting at 0x00000000) to consecutive locations starting at -0x80000000.  Not sure why it's negative though!
Then call the function with what looks like that address (without the -ve).

**Reply by kabiri** [reply]
the function FUN_00001420 is : 

```
undefined4 FUN_00001420(undefined4 param_1)
{
  return param_1;
}
```
I have the same problem
Why does the address decrease to such a large size?
And the function does not do much (ram address start from 40x0000000)

**Reply by Wall-AF** [reply]
You need to double check that the pseudo C code matches **exactly** the assembly (by hand if necessary).

**Reply by kabiri** [reply]
I know DOS assembly and I do not know much about micro assembly.
for example i dont know what means this move.l 

move.l     (0x0,A0,D2*0x4),D0
move.l     D0,(0x0,A1,D2*0x4)=>DAT_80000000
addq.l     #0x1,D2

**Reply by Wall-AF** [reply]
Mmm, looks like you've got some serious learning on your hands.  My assembly knowledge is limited to x86 (< 1999) and some other 1980s-90s stuff!
